### ISSUE TITLE
RabbitListener Add batch(), batchSize(), and receiveTimeout() Overrides

### ISSUE THREAD

**Issue by garyrussell** [spring-projects/spring-amqp#1545]
Similar to https://github.com/spring-projects/spring-kafka/issues/1872

Related to https://github.com/spring-projects/spring-amqp/issues/1544

**Comment by MartinFerret** [comment]
Hi @artembilan , I'd like to work on this, but I have a couple of questions first.

`batch()` was already added (3.0 per its Javadoc), so what remains here is `batchSize()` and `receiveTimeout()`. 

Both properties currently only exist on `SimpleRabbitListenerContainerFactory`, not on the abstract factory, so my plan would be to resolve them onto the endpoint (mirroring how `priority()` is parsed) and let the endpoint value win in `SimpleRabbitListenerContainerFactory.initializeContainer()`, the way `batchListener` is already handled.

Two questions before I start:

1. In #1544 you pointed to Spring Boot for configuring `receiveTimeout` via properties. I'd like to confirm the per-listener override on the annotation is still the direction you'd want here.

2. If it is, what should happen when these attributes are set on a listener bound to a `DirectRabbitListenerContainerFactory`, which has neither property? Silently ignore, or fail fast at startup?

Happy to open a PR once you confirm.

**Comment by artembilan** [comment]
Hey @MartinFerret !

I need some time to recall all of these.
We are busy at the moment with something else: will be back to you next week or so.

Thank you for the help!

**Comment by MartinFerret** [comment]
Of course, no worries, take your time!

**Comment by artembilan** [comment]
@MartinFerret ,

Yes, let's go with those new annotation properties!
But let's call it `batchReceiveTimeout` as it is called in the target listener container for that batch-specific operation.
For example in the `RabbitAmqpListenerContainer`:
```
	/**
	 * Set a timeout in milliseconds for how long a batch gathering process should go.
	 * Therefore, the batch is released as a single message whatever first happens:
	 * this timeout or {@link #setBatchSize(int)}.
	 * Default 30 seconds.
	 * @param batchReceiveTimeout the timeout for gathering a batch.
	 */
	public void setBatchReceiveTimeout(long batchReceiveTimeout) {
```
I don't think we need to pull these new properties to the parent class of the `SimpleRabbitListenerContainerFactory` and pollute environment for the `DirectRabbitListenerContainerFactory`.
Then we would need to copy/paste the logic for these new properties into the `RabbitAmqpListenerContainerFactory`.

I think all the unknown properties from the annotation are just ignored in the `DirectRabbitListenerContainerFactory`.
So, should not be a harm to have a couple more ignored 😄

**Comment by MartinFerret** [comment]
One semantics question before I go further: on SimpleMessageListenerContainer, batchSize only drives consumer-side batching when consumerBatchEnabled is true, and batchReceiveTimeout only applies in that mode, but consumerBatchEnabled stays at the factory level. 

So @RabbitListener(batch = "true", batchSize = "10") on a factory without consumerBatchEnabled would silently not produce batches of 10. 

Do you want a pure override (document the caveat), or should batch = "true" on the annotation also imply consumer batching?

**Comment by artembilan** [comment]
I think that is documented some way already: https://docs.spring.io/spring-amqp/reference/amqp/receiving-messages/batch.html

> Spring Boot provides a configuration property for consumerBatchEnabled and batchSize, but not for batchListener. Starting with version 3.0, setting consumerBatchEnabled to true on the container factory also sets batchListener to true. When consumerBatchEnabled is true, the listener must be a batch listener.

**Comment by MartinFerret** [comment]
Right, I'll keep the annotation attributes as pure value overrides then (no implied consumerBatchEnabled), consistent with what's already documented, and I'll make that explicit in batch.adoc so the interaction is clear from the annotation side too. 

Opening the PR shortly.
